horovod/common/mpi_context.cc

```diff
@@ -77,5 +85,126 @@ int MPIContext::GetMPITypeSize(DataType dtype) {
   return out;
 }
 
+#ifdef HAVE_DDL
+void MPIContext::Initialize(const std::vector<int>& ranks,
+                            DDLContext& ddl_context, CUDAContext& cuda_context)
+#else
+void MPIContext::Initialize(const std::vector<int>& ranks)
+#endif
+{
+  // Initialize MPI if it was not initialized. This must happen on the
+  // background thread, since not all MPI implementations support being called
+  // from multiple threads.
+  //
+  // In some cases MPI library has multi-threading support, but it slows down
+  // certain components, e.g. OpenIB BTL in OpenMPI gets disabled if
+  // MPI_THREAD_MULTIPLE is requested.
+  //
+  // By default, we will ask for multiple threads, so other libraries like
+  // mpi4py can be used together with Horovod if multi-threaded MPI is
+  // installed.
+  auto mpi_threads_disable = std::getenv(HOROVOD_MPI_THREADS_DISABLE);
+  int required = MPI_THREAD_MULTIPLE;
+  if (mpi_threads_disable != nullptr &&
+      std::strtol(mpi_threads_disable, nullptr, 10) > 0) {
+    required = MPI_THREAD_SINGLE;
+  }
+  int provided;
+  int is_mpi_initialized = 0;
+  MPI_Initialized(&is_mpi_initialized);
+  if (is_mpi_initialized) {
+    MPI_Query_thread(&provided);
+    if (provided < MPI_THREAD_MULTIPLE) {
+      LOG(WARNING)
+          << "MPI has already been initialized without "
+             "multi-threading support (MPI_THREAD_MULTIPLE). This will "
+             "likely cause a segmentation fault.";
+    }
+  } else {
+#if HAVE_DDL
+    // DDL comes with IBM Spectrum MPI
+    // and needs to initialize MPI with the proper license.
+    DDLAllreduce::DDLInit(&ddl_context, &cuda_context);
+#else
+    MPI_Init_thread(nullptr, nullptr, required, &provided);
```

Comment:
```
class MPIContextManager:
public:
    void Initialize(required, provided);

    void Finalize();

class DDLContextManager : public MPIContextManager
public:
    DDLContextManager(ddl_context, cuda_context)

class MPIContext:
public:
    void Initialize(ranks, context_manager);
```

Reply:
We are currently using the above solution to decouple DDL from MPI context.